core/rfb.js

```diff
@@ -1496,6 +1500,99 @@ export default class RFB extends EventTargetMixin {
         return true;
     }
 
+    _negotiateARDAuth() {
+
+        function stringToBytes(str) {
+            let bytes = new Uint8Array(str.length);
+            for (let i = 0; i < str.length; ++i) {
+                bytes[i] = str.charCodeAt(i);
+            }
+            return bytes;
+        }
+
+        function bytesToHex(byteArray) {
+            return Array.from(byteArray, byte => ('0' + (byte & 0xFF).toString(16)).slice(-2)).join('');
+        }
+
+        function hexToBytes(hex) {
+            if (hex.length%2 != 0) hex = "0"+hex;
+            let bytes = [];
+            for (let c = 0; c < hex.length; c += 2) {
+                bytes.push(parseInt(hex.substr(c, 2), 16));
+            }
+            return bytes;
+        }
+
+        if (this._rfbCredentials.username === undefined ||
+            this._rfbCredentials.password === undefined) {
+            this.dispatchEvent(new CustomEvent(
+                "credentialsrequired",
+                { detail: { types: ["username", "password"] } }));
+            return false;
+        }
+
+        if (this._sock.rQwait("read ard", 4)) { return false; }
+
+        let generator = this._sock.rQshiftBytes(2);   // DH base generator value
+
+        let keyLength = this._sock.rQshift16();
+
+        if (this._sock.rQwait("read ard keylength", keyLength*2, 4)) { return false; }
+
+        let primeBytes = this._sock.rQshiftBytes(keyLength);  // predetermined prime modulus
+
+        let serverPubKeyBytes = this._sock.rQshiftBytes(keyLength); // other party's public key
+
+        let prime = BigInt("0x"+bytesToHex(primeBytes));
+
+        let clientPivKeyBytes = window.crypto.getRandomValues(new Uint8Array(keyLength));
+        let clientPrivKey = BigInt("0x"+bytesToHex(clientPivKeyBytes));
+
+        let ghd = BigInt("0x"+bytesToHex(generator));
+
+        // calculate the DH keys
+        let clientPubKey = modPow(ghd, clientPrivKey, prime);
+
+        let serverPubKey = BigInt("0x"+bytesToHex(serverPubKeyBytes));
+
+        let sharedKey = modPow(serverPubKey, clientPrivKey, prime);
+
+        // encrypt the credentials
+        function encodeUtf8(str) {
+            return unescape(encodeURIComponent(str));
+        }
+
+        let username = encodeUtf8(this._rfbCredentials.username).substring(0, 63);
+        let password = encodeUtf8(this._rfbCredentials.password).substring(0, 63);
+
+        let paddedUsername = username + '\0' + bytesToHex(window.crypto.getRandomValues(new Uint8Array(34))).substring(0, 63);
+        let paddedPassword = password + '\0' + bytesToHex(window.crypto.getRandomValues(new Uint8Array(34))).substring(0, 63);
+
+        let credentials = paddedUsername.substring(0, 64) + paddedPassword.substring(0, 64);
+
+        let sharedKeyString = Array.from(hexToBytes(sharedKey.toString(16)), byte => String.fromCharCode(byte)).join('');
+
+        (async () => {
```

Comment:
Using async functions is okay since it is also well supported by all current browsers.

However this is probably not the correct way to do so. Calling an async function will schedule it to complete at a later time. So you will be changing `this._rfbInitState` before you've completed the encryption and sent the data.

This happens to work anyway since `RFB` will go back to waiting for more data from the server, which it won't get since the server is waiting for data from the async function. But this is fragile, so we need to get this a bit more robust.

The proper solution would be to rewrite all protocol handling to be async. But that's a massive change, so not suitable here.

What you need to do is to make sure that the caller knows you are waiting for something. So you return `false` once you've kicked off the async function. You then need to set up a couple of things:

1. Some tracking that you are waiting for the async function, so if `_negotiateARDAuth()` gets called again you can throw an error.
2. A resolve handler that updates `this._rfbInitState` and calls `this._initMsg()`. Compare to what happens when we need to wait for credentials, and then later `sendCredentials()` gets called to resume things.

All of this should ideally also have unit tests to make sure nothing breaks in the future.